arangod/Aql/UnsortingGatherExecutor.cpp

```diff
@@ -0,0 +1,103 @@
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Tobias Gödderz
+////////////////////////////////////////////////////////////////////////////////
+
+#include "UnsortingGatherExecutor.h"
+#include <Logger/LogMacros.h>
+
+#include "Aql/IdExecutor.h"  // for IdExecutorInfos
+#include "Aql/MultiDependencySingleRowFetcher.h"
+#include "Aql/OutputAqlItemRow.h"
+#include "Aql/Stats.h"
+#include "Basics/Exceptions.h"
+#include "Basics/debugging.h"
+#include "Basics/voc-errors.h"
+
+using namespace arangodb;
+using namespace arangodb::aql;
+
+UnsortingGatherExecutor::UnsortingGatherExecutor(Fetcher& fetcher, Infos& infos)
+    : _fetcher(fetcher) {}
+
+UnsortingGatherExecutor::~UnsortingGatherExecutor() = default;
+
+auto UnsortingGatherExecutor::produceRows(OutputAqlItemRow& output)
+    -> std::pair<ExecutionState, Stats> {
+  while (!output.isFull() && !done()) {
+    // Note that fetchNextRow may return DONE (because the current dependency is
+    // DONE), and also return an unitialized row in that case, but we are not
+    // DONE completely - that's what `done()` is for.
+    auto [state, inputRow] = fetchNextRow(output.numRowsLeft());
+    if (state == ExecutionState::WAITING) {
+      return {state, {}};
+    }
+    // HASMORE => inputRow.isInitialized()
+    TRI_ASSERT(state == ExecutionState::DONE || inputRow.isInitialized());
+    if (inputRow.isInitialized()) {
+      output.copyRow(inputRow);
+      TRI_ASSERT(output.produced());
+      output.advanceRow();
+    }
+  }
+
+  auto state = done() ? ExecutionState::DONE : ExecutionState::HASMORE;
+  return {state, {}};
+}
+
+auto UnsortingGatherExecutor::fetcher() const noexcept -> const Fetcher& {
+  return _fetcher;
+}
+
+auto UnsortingGatherExecutor::fetcher() noexcept -> Fetcher& {
+  return _fetcher;
+}
+
+auto UnsortingGatherExecutor::numDependencies() const
+    noexcept(noexcept(_fetcher.numberDependencies())) -> size_t {
+  return _fetcher.numberDependencies();
+}
+
+auto UnsortingGatherExecutor::fetchNextRow(size_t atMost)
+    -> std::pair<ExecutionState, InputAqlItemRow> {
+  auto res = fetcher().fetchRowForDependency(currentDependency(), atMost);
+  if (res.first == ExecutionState::DONE) {
+    advanceDependency();
+  }
+  return res;
+}
+
+auto UnsortingGatherExecutor::done() const noexcept -> bool {
+  return _currentDependency >= numDependencies();
```

Comment:
So I assume the executor tries to fetch *all* results from the first dependency, then all results from the second, until it has reached the last one? This should be in the description I asked for.

I stumbled over this because I thought "what happens if one dependency has said "wait for me" and we decided to check the next one, just in case it has stuff ready already. I am hence further assuming that this is something for another bit of the code or another executor to do?

Reply:
Yes, the `UnsortedGatherExecutor` fires requests on all its dependencies in parallel. This one purposefully works only in a purely synchronous fashion, though the specific order in which the dependencies are pulled does not matter.

Reply:
What is the purpose of doing this in a synchronous fashion? The comment now says that one should not rely on the order in which the dependencies are read.

Reply:
That's correct. The `UnsortedGatherExecutor` works on its dependencies in parallel, i.e., if it gets a WAITING, it pulls the next one. This is only feasible in certain situations where it can't lead to two overlapping requests on the same DBServer. For the other cases, the `UnsortingGatherExecutor` is used. So with "synchronous" I tried to discriminate between the "asynchronous" `UnsortedGatherExecutor`. Perhaps the wording isn't chosen well, I can rewrite or extend the comment if you think that would help.